### ISSUE TITLE
Issue running zipped release

### ISSUE THREAD

**Issue by Parasrah** [Parasrah/ddb-proxy#14]
From a report on the `ddb-importer` regarding this repo:

> Hi Brad,
> 
> I waited until this afternoon to start in with some testing. I just downloaded release v0.4.0 and attempted to run it on the command line, in the folder I unzipped it into with the following command:
> 
> bin/ddb_proxy start
> 
> and I received the following error:
> 
> Kenneths-MacBook-Pro:ddb-proxy kemarks$ bin/ddb_proxy start
> bin/ddb_proxy: /Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/Modules/ddb-proxy/releases/0.3.0/elixir: /nix/store/9ywr69qi622lrmx5nn88gk8jpmihy0dz-bash-4.4-p23/bin/sh: bad interpreter: No such file or directory
> bin/ddb_proxy: line 57: /Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/Modules/ddb-proxy/releases/0.3.0/elixir: Undefined error: 0
> 
> Do I need to install something?
> 
> I am running on a MacBook Pro under Mojave 10.14.6 and I did set all the environment variables in my .bash_profile.
> 
> Thanks again for your work on the development of this and for letting me do some testing on it.
> 
> - Ken

**Comment by Parasrah** [comment]
This is very interesting. I'm assuming you don't use the `nix` package manager @KennethEMarks?

If not, it's probably because I used `nix` to build the package. I'll explain a little bit for anyone that is interested, but please feel free to the TLDR at the end if you're not.

## Long explanation

Most package managers in `unix` environments mutate your global environment. For example, `homebrew`, `apt-get`, `pacman`, etc. When you install a package using these package managers, they all put executable binaries in the same places. Such as `/bin`, `/usr/bin`, etc. That means if you want to install two versions of software, you can't because there can only be one `/bin/bash` for example.

Nix changes that by installing everything to `/nix/store/<hash>-package-<version>`. This way you could have a `/nix/store/9ywr69qi622lrmx5nn88gk8jpmihy0dz-bash-4.4-p23/bin/bash` for `bash` 4.4, and a `/nix/store/f2ca1bb6c7e96dafe4687e6ccc26dfd2-bash-4.3/bin/bash` for `bash` 4.3, and they are both installed on your system at the same time. Whichever one comes first in your `$PATH` is the one that gets picked up in a shell, so you can dynamically switch between the two whenever you want. This isn't particularly useful for `bash`, but is for things like `python`, `node`, `elixir` and stuff like that.

Issue is, a lot of scripts and programs are written to look for `sh` or `bash` in `/bin/bash`, so to make sure that things are consistent, packages installed using `nix` are modified by the package manager to point at the nix store for these things instead of `/bin/`. And given that elixir releases bundle everything into the release needed to run the release, it's probably also bundling in the modifications the package manager made to my `elixir` install. This also applies to my CI responsible for building the release, because it's using `nix` to install the dependencies.

I didn't realize this was an issue, and it's definitely something I will need to sort out. For this project it should be easy enough to change the CI to build using a version of elixir installed via `apt-get` instead, but it raises a larger problem with my elixir/nix workflow I will need to sort out.

## TLDR

I will have to change the CI pipeline to not use the `nix` package manager to install elixir. I won't be able to do that tonight, but will make that my next priority

**Comment by Parasrah** [comment]
Side note for you @KennethEMarks, if you want to avoid setting environment variables in your `.bashrc` you can also prepend them to the command. Such as `MY_ENV='http://localhost/test ./bin/ddb_proxy start`. If you would like to have them set in your `.bashrc` so it's kind of set and forget, I could change the variables to something more unique like `DDB_TOKEN_API` etc for example? Just to make sure it never conflicts with something else or overrides something for another program you might be running

**Comment by KennethEMarks** [comment]
Hi Brad,

I do not use the nix package manager. I typically use homebrew on my Mac to install packages. If you think I should install it, let me know. Otherwise I’ll await further instructions on how to proceed.

Best regards,

- Ken

**Comment by KennethEMarks** [comment]
Hi Brad,

I do think it would be a good idea to preface the environment variables with DDB_ as that gives maximum flexibility to users if they want to put it in their .bashrc (or .bash_profile), or prepend them to the command.

- Ken

**Comment by Parasrah** [comment]
@KennethEMarks if you do install it there is the chance it would unblock you in the short term, but I hopefully will have a chance tonight to revise my CI to build releases that should work outside of the `nix` ecosystem.

**Comment by Parasrah** [comment]
Hey @KennethEMarks, I've build a version `0.4.1` now that was built without `nix`. I'm afraid I can't test it on my machine, I use NixOS and the release doesn't work for me. Hopefully it works for you though!

**Comment by Parasrah** [comment]
Oops didn't mean to close this, referenced the wrong bug

**Comment by KennethEMarks** [comment]
Hi Brad,

Do you want me to test 0.4.1 or 0.5.2?

- Ken

**Comment by KennethEMarks** [comment]
Hi Brad,

I downloaded release 0.5.2 and executed the following command on the command line in the ddb-proxy directory and got the following error:

bin/ddb-proxy start

Kenneths-MacBook-Pro:ddb-proxy kemarks$ bin/ddb_proxy start
/Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/proxy/ddb-proxy/releases/0.5.1/../../erts-10.6/bin/erl: line 12: /Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/proxy/ddb-proxy/erts-10.6/bin/erlexec: cannot execute binary file
/Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/proxy/ddb-proxy/releases/0.5.1/../../erts-10.6/bin/erl: line 12: /Users/kemarks/Documents/Development/Gaming/DnD/FoundryVTT/proxy/ddb-proxy/erts-10.6/bin/erlexec: Undefined error: 0
Kenneths-MacBook-Pro:ddb-proxy kemarks$ 

I did change my environment variables to reflect the new DDB_ prefixes as well.

- Ken

**Comment by Parasrah** [comment]
Hmm that was the same error I got, maybe it should be working on my machine too. I should have some time this weekend to dig into it more. An yeah any of the latest releases should have been okay, I fixed some other things as well

**Comment by KennethEMarks** [comment]
Cool,

I’ll poke around some more this weekend with it.

- Ken

**Comment by Parasrah** [comment]
It would seem I misunderstood the flexibility of elixir releases, as I neglected to read the first sentence in the documentation

> Once a release is assembled, it can be packaged and deployed to a target, as long as the target runs on the same operating system (OS) distribution and version as the machine running the mix release command.

This is obviously not great, and I'm going to have to look into alternatives. Had I known this before I may have written it in python or something else widely available on most OS's, but it's too late for that unfortunately. Right now I can think of a couple alternatives:

1) instruct users to install elixir themselves and run the server using it (similar to it being nodejs)
2) instruct users to install nix, so they can easily package their own release

At this point I'm tempted to go with route (2), as I think it's easier and I'm more confident it would work reliably (that is kind of the whole point of nix). The drawback of that would be leaving windows users out in the cold (nix is only supported on unix systems like Linux or MacOS)

I could also provide documentation for both and recommend people take route (1) unless they are windows and aren't using WSL.

In any case I will give this some more thought and hopefully act on it tomorrow given the opportunity. Sorry for any time this wasted for you testing the release

**Comment by Parasrah** [comment]
I've updated the build process to make it significantly easier for someone to perform themselves, and in the case of MacOS & Linux you should be able to build your own release using `nix` now. I've updated the `README` to reflect this and provide alternate instructions for windows users

**Comment by Parasrah** [comment]
Closing this in the short term as there are no longer zipped releases, but if there are any issues with the new process please open an issue. Hope it works for you!
